android/gltfio-android/src/main/cpp/AssetLoader.cpp

```diff
@@ -23,18 +23,151 @@
 
 #include <gltfio/AssetLoader.h>
 #include <gltfio/MaterialProvider.h>
+#include <utils/debug.h>
 
 #include "common/NioUtils.h"
 
+#include "MaterialKey.h"
+
 using namespace filament;
 using namespace gltfio;
 using namespace utils;
 
+class JavaMaterialProvider : public MaterialProvider {
+    JNIEnv* const mEnv;
+    const jobject mJavaProvider;
+    mutable Material** mPreviousMaterials = nullptr;
+
+    jclass mMaterialKeyClass;
+
+    jmethodID mMaterialKeyConstructor;
+    jmethodID mCreateMaterialInstance;
+    jmethodID mGetMaterials;
+    jmethodID mNeedsDummyData;
+    jmethodID mDestroyMaterials;
+    jmethodID mMaterialInstanceGetNativeObject;
+    jmethodID mMaterialGetNativeObject;
+
+public:
+    JavaMaterialProvider(JNIEnv* env, jobject provider) : mEnv(env),
+            mJavaProvider(env->NewGlobalRef(provider)) {
+        mMaterialKeyClass = env->FindClass(JAVA_MATERIAL_KEY);
+        mMaterialKeyClass = (jclass) env->NewGlobalRef(mMaterialKeyClass);
+        assert_invariant(mMaterialKeyClass);
+
+        mMaterialKeyConstructor = env->GetMethodID(mMaterialKeyClass, "<init>", "()V");
+        assert_invariant(mMaterialKeyConstructor);
+
+        jclass materialInstanceClass = env->FindClass("com/google/android/filament/MaterialInstance");
+        mMaterialInstanceGetNativeObject = env->GetMethodID(materialInstanceClass, "getNativeObject", "()J");
+        assert_invariant(mMaterialInstanceGetNativeObject);
+
+        jclass materialClass = env->FindClass("com/google/android/filament/Material");
+        mMaterialGetNativeObject = env->GetMethodID(materialClass, "getNativeObject", "()J");
+        assert_invariant(mMaterialGetNativeObject);
+
+        jclass providerClass = env->GetObjectClass(provider);
+
+        mCreateMaterialInstance = env->GetMethodID(providerClass, "createMaterialInstance",
+                "(L" JAVA_MATERIAL_KEY ";[ILjava/lang/String;)Lcom/google/android/filament/MaterialInstance;");
+        assert_invariant(mCreateMaterialInstance);
+
+        mGetMaterials = env->GetMethodID(providerClass, "getMaterials",
+                "()[Lcom/google/android/filament/Material;");
+        assert_invariant(mGetMaterials);
+
+        mNeedsDummyData = env->GetMethodID(providerClass, "needsDummyData","(I)Z");
+        assert_invariant(mNeedsDummyData);
+
+        mDestroyMaterials = env->GetMethodID(providerClass, "destroyMaterials", "()V");
+        assert_invariant(mDestroyMaterials);
+    }
+
+    ~JavaMaterialProvider() override {
+        mEnv->DeleteGlobalRef(mMaterialKeyClass);
+        mEnv->DeleteGlobalRef(mJavaProvider);
+        delete mPreviousMaterials;
+    }
+
+    MaterialInstance* createMaterialInstance(MaterialKey* config, UvMap* uvmap, const char* label) override {
```

Comment:
Is it not possible to call up to Java, and have all that java stuff be donne in Java?

Reply:
This is called from the native code.  The point of this PR is to allow Java clients to write their own MaterialProvider class.

Reply:
yes but could use have a java method that's like "doAllTheJavaStuff()" that you call from native and does all that?

Reply:
In this particular case, this idea would not result in an simplification, most of the work is copying between JNI and native.